AnkiHubSoftware/ankihub_addon
# [BUILD-451] feat: Handle message from iframe to open Anki browser

The AnkiHub AI UI will have buttons that, when clicked, will open the Anki browser with specific notes as search results. This task will implement the Anki-side for that.

The AnkiHub AI templates will be able to use java script like this to send a message to python:
```js
const message = {
    // The format of the message is "<command> <args_dict>"
    message: `ankihub_open_browser {
      "noteIds": [
        "00000000-0000-0000-0000-000000000001", "00000000-0000-0000-0000-000000000002"
      ]
    }`,
    sendToPython: true
};
window.parent.postMessage(message, '*');
```

The reviewer javascript will receive this message, see that `sendToPython` is true and use `pycmd` to send the message to Python. (We can only send strings to Python.)
The message handler in Python will get this message and parse it into the name of the command (ankihub_open_browser) and the arguments (provided as json dict).
Then it will open the Anki Browser with the provided notes as search results.

## Related issues
https://ankihub.atlassian.net/jira/software/c/projects/BUILD/boards/1?selectedIssue=BUILD-451

## Proposed changes
- The javascript in the reviewer webview receives messages sent using `postMessage` from the AnkiHub AI iframe and relays them to Python using the python-js bridge.
- Python reacts to the `ankihub_open_browser` message (sent via the python-js bridge) by opening the browser. 
- Add tests

## How to reproduce
- Comment out this code to make the chatbot show up for any AnkiHub deck:
  https://github.com/ankipalace/ankihub_addon/blob/1a8b0eb4b3c2b030b0c538c44fea8fc6014e151f/ankihub/gui/reviewer.py#L104-L107
- Add this code:
  ```js
  const message = {
    message: `ankihub_open_browser {"noteIds": [
      "59d5f5c4-2d52-413b-8895-5999c702200a", "d3ec9513-b4d6-43ab-b5dc-3a9189a93d61"
    ]}`,
    sendToPython: true
  }
  window.parent.postMessage(message, '*')
  ```
  to the web app `chatbot.html` template below this:
  https://github.com/ankipalace/ankihub/blob/f6c9d7fce05f46a03fcdce86b5f624b449496a1c/ankihub/templates/ai/chatbot.html#L92-L93
  This way the `ankihub_open_browser` message will be sent when you send a message to the chatbot. (This is just for testing and not how it will work for real.)
- Enable the `chatbot` feature flag on your local ankihub web app
- Open an ankihub note in the reviewer
- Follow the steps in the video
  - You won't have the same note ids as me, so you will get no search results in the browser. You will still be able to see that the search query shows the note ids provided above.  

## Screenshots and videos

https://github.com/ankipalace/ankihub_addon/assets/31575114/4e54b2c9-6d58-47c4-872b-146c760dc1df
